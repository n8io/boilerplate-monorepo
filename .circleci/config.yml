version: 2.1

jobs:
  test:
    docker:
      - image: circleci/node:12.13.0
    environment:
      TERM: xterm
    working_directory: ~/app
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-deps
            - v1-deps-{{ .Branch }}
            - v1-deps-{{ .Branch }}-{{ checksum "package.json" }}
      - run:
          name: Installing dependencies
          command: yarn install --frozen-lockfile
      - run:
          name: Running linters
          command: yarn lint
      - run:
          name: Build common dependencies
          command: yarn -s build:common && yarn -s build:ui-common
      - run:
          name: Running unit tests
          command: yarn test:ci
      - save_cache:
          key: v1-deps-{{ .Branch }}-{{ checksum "package.json" }}
          paths:
            - node_modules
            - ~/.cache
      - store_test_results:
          path: .tmp/test-results

  build:
    docker:
      - image: circleci/node:12.13.0
    environment:
      TERM: xterm
    working_directory: ~/app
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-deps
            - v1-deps-{{ .Branch }}
            - v1-deps-{{ .Branch }}-{{ checksum "package.json" }}
      - run:
          name: Installing dependencies
          command: yarn install --frozen-lockfile
      - run:
          name: Build all the things
          command: yarn -s build

  deploy:
    docker:
      - image: circleci/node:12.13.0
    steps:
      - checkout
      - run:
          name: Deploy to Heroku
          command: |
            git push https://heroku:$HEROKU_API_KEY@git.heroku.com/$HEROKU_APP_NAME.git master

workflows:
  version: 2
  features:
    jobs:
      - test
      - build:
          requires:
            - test
          filters:
            branches:
              only:
                - master
      - approve-deployment:
          type: approval
          requires:
            - build
      - deploy:
          requires:
            - approve-deployment
          filters:
            branches:
              only:
                - master
