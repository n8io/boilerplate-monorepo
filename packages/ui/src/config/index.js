import { LocalStorage } from '@boilerplate-monorepo/ui-common';
import { defaultTo, filter, fromPairs, keys, map, pipe } from 'ramda';

const DEVELOPMENT_PREFIX = 'dev';
const PRODUCTION_PREFIX = 'prod';

// eslint-disable-next-line no-process-env
const { NODE_ENV = PRODUCTION_PREFIX, ...envVars } = process.env;

const REACT_APP_PREFIX = 'REACT_APP_';

const reactVars = pipe(
  keys,
  filter(key => key.toString().startsWith(REACT_APP_PREFIX)),
  map(key => [key.replace(REACT_APP_PREFIX, ''), envVars[key]]),
  fromPairs,
  ({ PUBLIC_URL, SENTRY_ENV, ...rest }) => ({
    ...rest,
    PUBLIC_URL: defaultTo('/', PUBLIC_URL),
    SENTRY_ENV: defaultTo(NODE_ENV, SENTRY_ENV),
  })
)(envVars);

const config = {
  ...reactVars,
  NODE_ENV,
  isDebug: Boolean(localStorage.getItem(LocalStorage.DEBUG)),
  isDevelopment: NODE_ENV.startsWith(DEVELOPMENT_PREFIX),
  isProduction: NODE_ENV.startsWith(PRODUCTION_PREFIX),
};

export { config };
