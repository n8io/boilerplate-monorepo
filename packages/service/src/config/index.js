import { Utils } from '@boilerplate-monorepo/common';
import 'dotenv/config';
import {
  always,
  both,
  complement,
  cond,
  defaultTo,
  either,
  evolve,
  includes,
  isEmpty,
  mergeRight,
  pick,
  pipe,
  prop,
  propOr,
  startsWith,
  T,
  toLower,
} from 'ramda';
import { ProcessEnvKeys } from 'types/processEnvKeys';

// eslint-disable-next-line no-process-env
const raw = pick(ProcessEnvKeys.values, process.env);

const defaults = {
  ACCESS_TOKEN_EXPIRY: '5m',
  ACCESS_TOKEN_SECRET: 'auth-token-secret',
  CAPTCHA_SECRET: '0x0000000000000000000000000000000000000000',
  CI: false,
  DATABASE_MIGRATION_SCHEMA: 'public',
  DATABASE_MIGRATION_TABLE_NAME: 'migrations',
  DATABASE_SCHEMA: 'main',
  DATABASE_URL: '',
  DATABASE_USE_SSL: false,
  DEBUG: `${raw.npm_package_name}*`,
  EMAIL_FROM_ADDRESS: `noreply@local.host`,
  EMAIL_FROM_NAME: 'ðŸ¤– Automated Robot',
  EMAIL_SMTP_TIMEOUT_SECONDS: 3,
  EMAIL_SMTP_URL: '',
  ENGINE_API_KEY: '',
  ENGINE_SCHEMA_TAG: 'local',
  HTTPS: true,
  NODE_ENV: 'development',
  PORT: 4000,
  REDIS_KEY_PREFIX: 'development',
  REDIS_URL: '',
  REFRESH_TOKEN_EXPIRY: '7d',
  REFRESH_TOKEN_SECRET: 'refresh-token-secret',
  SENTRY_DSN: null,
  SHOW_CONFIG: false,
  UI_HOST_URI: 'https://local.host:3000',
};

const config = mergeRight(defaults, raw);

const normalize = pipe(prop('NODE_ENV'), defaultTo('development'), toLower);

const isTest = (env) =>
  includes(env, ['ci', 'jest', 'test']) ||
  (typeof describe !== 'undefined' && typeof test !== 'undefined');

const isDevelopment = pipe(
  normalize,
  either(startsWith('dev'), startsWith('local'))
);

const isStaging = pipe(normalize, startsWith('st'));
const isProduction = pipe(normalize, startsWith('pr'));

const toEnvironment = pipe(
  normalize,
  cond([
    [isTest, always('test')],
    [isProduction, always('production')],
    [isStaging, always('staging')],
    [T, always('development')],
  ])
);

const isSqlDebug = always(
  pipe(propOr('', ProcessEnvKeys.DEBUG), includes('sql'))
)(config);

const isUndefined = (value) => typeof value === 'undefined';
const isUndefinedOrEmpty = either(isUndefined, isEmpty);
const isNotDevAndEmpty = both(complement(isDevelopment), isUndefinedOrEmpty);

const isTelemetryEnabled = pipe(
  propOr(false, ProcessEnvKeys.SENTRY_DSN),
  Boolean
);

const requiredEnvVars = [
  { failureTest: isUndefinedOrEmpty, name: ProcessEnvKeys.DATABASE_URL },
  { failureTest: isNotDevAndEmpty, name: ProcessEnvKeys.REDIS_URL },
];

const missingEnvVars = requiredEnvVars.filter(({ name, failureTest }) =>
  failureTest(config[name])
);

if (isProduction(config) && missingEnvVars.length) {
  const messages = missingEnvVars.map(
    ({ name }) =>
      `${name} was not set. The application will not function correctly without it.`
  );

  // eslint-disable-next-line no-console
  console.error(messages.join('\n'));

  // eslint-disable-next-line no-process-exit
  process.exit(1);
}

const merged = pipe(
  evolve({
    DATABASE_USE_SSL: Utils.toBool,
    EMAIL_SMTP_TIMEOUT_SECONDS: Utils.toBool,
    HTTPS: Utils.toBool,
    PORT: Utils.toNumber,
    REDIS_KEY_PREFIX: defaultTo(config.NODE_ENV),
    SHOW_CONFIG: Utils.toBool,
  }),
  Utils.renameKeys({
    [ProcessEnvKeys.npm_package_name]: 'name',
    [ProcessEnvKeys.npm_package_version]: 'version',
  }),
  mergeRight({
    environment: toEnvironment(config),
    isDevelopment: isDevelopment(config),
    isProduction: isProduction(config),
    isSqlDebug: isSqlDebug(),
    isStaging: isStaging(config),
    isTelemetryEnabled: isTelemetryEnabled(config),
    isTest: isTest(config),
  })
)(config);

// eslint-disable-next-line no-console
merged.SHOW_CONFIG && console.log(JSON.stringify(merged, null, 2));

export { merged as config };
