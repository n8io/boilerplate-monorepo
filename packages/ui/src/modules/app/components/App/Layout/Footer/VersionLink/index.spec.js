import { render } from '@testing-library/react';
import * as Config from 'config';
import React from 'react';
import { VersionLink } from '.';

const renderComponent = () => render(<VersionLink />);

describe('<VersionLink />', () => {
  describe('when there is a release', () => {
    beforeEach(() => {
      td.replace(Config, 'config', {
        RELEASE: 'RELEASE',
      });
    });

    test('renders the release properly', () => {
      expect(renderComponent().container.firstChild).toMatchSnapshot();
    });
  });

  describe('when there is a commit hash', () => {
    beforeEach(() => {
      td.replace(Config, 'config', {
        COMMIT_HASH: 'COMMIT_HASH',
        RELEASE: 'RELEASE',
      });
    });

    test('renders the release properly', () => {
      expect(renderComponent().container.firstChild).toMatchSnapshot();
    });
  });
});
