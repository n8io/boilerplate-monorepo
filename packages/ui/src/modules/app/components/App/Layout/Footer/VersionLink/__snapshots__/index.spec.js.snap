// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VersionLink /> when there is a commit hash renders the release properly 1`] = `
<a
  class="ExternalLink__StyledLink-dCpmsY cMdWQu"
  href="https://github.com/n8io/boilerplate-monorepo/commit/COMMIT_HASH"
  rel="noopener noreferrer"
  target="_external"
>
  COMMIT_HASH
</a>
`;

exports[`<VersionLink /> when there is a release renders the release properly 1`] = `
<a
  class="ExternalLink__StyledLink-dCpmsY cMdWQu"
  href="https://github.com/n8io/boilerplate-monorepo/releases/tag/RELEASE"
  rel="noopener noreferrer"
  target="_external"
>
  RELEASE
</a>
`;
